### PR TITLE
`General`: Fix issues with tutor performance in assessment dashboard

### PR DESCRIPTION
**Pull request by Olimbia23**
#### General

- [X] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] *This is a small issue that I tested locally and was confirmed by another developer on a test server
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
#### Server
- [ ] *I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [ ] *I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] *I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [ ] *I implemented the changes with a good performance and prevented too many database calls.
- [ ] *I documented the Java code using JavaDoc style.
#### Client
- [X] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [X] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [ ] *I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [X] I documented the TypeScript code using JSDoc style.
- [X] I added multiple screenshots/screencasts of my UI changes.
- [X] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ] *I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] *I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

Instructor is interested in cases when a particular tutor performs worse than the average of the group. It would be nice to see which tutors have issues with performance and what exactly is wrong: bad scoring, bad rating or a lot of complaints? This PR solves this by computing the issues with tutor performance and showing it as a list in course assessment dashboard as well as marking the tutors with bad performance in the tutor leaderboard with ⚠️ near their name.

### Description

Issues with tutor performance are computed on the fly the client side. `TutorIssueRatingChecker`, `TutorIssueScoreChecker`, `TutorIssueComplaintsChecker` check if tutor has a performance issue with rating, score or complaints.
When it comes to complaints, complaints per submission ratio is computed, this way the performance issue is computed fairly.

Tutor performance issue is computed by computing the course average +/- 20%. This is the threshold value. If the tutor surpasses it, then he/she has a performance issue, which will be shown in the list.

### Steps for Testing

1. Log in to Artemis as Instructor
2. Navigate to Course Administration
3. Navigate to Course Assessment Dashboard
4. You should see the list with issues with tutor performance

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Before
before

#### After - logged in as Instructor
Screenshot 2021-10-22 at 3 49 47 PM

### CODE REVIEW

**Review by julian-christl — DISMISSED** [review]
Code looks good so far, let's monitor the values and see how they behave in production 👍

**Review by Olimbia23 — COMMENTED** [review]
Thank you for your review @lschlesinger.
I applied all your suggestions 😇

File: src/main/webapp/app/course/dashboards/assessment-dashboard/tutor-issue.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        public tutorName: string,
+        public numberOfTutorItems: number,
+        public averageTutorValue: number,
+        public allowedRange: Range,
+        public translationKey: string,
+    ) {}
+}
+
+/**
+ * Represents the an inclusive range with a lower and upper bound
```
Comment (lschlesinger) [inline src/main/webapp/app/course/dashboards/assessment-dashboard/tutor-issue.ts:15]: the + an, one could be removed

**Review by lschlesinger — CHANGES_REQUESTED** [review]
There is  a failing client test suite for `TutorLeaderboardComponent`
https://github.com/ls1intum/Artemis/runs/4052298509?check_suite_focus=true

Once this is fixed, I can manually test again :)

**Review by lschlesinger — DISMISSED** [review]
I tested again on TS1. issues were shown as expected 👍

File: src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts [hunk 2]
```diff
@@ -143,85 +146,129 @@ describe('AssessmentDashboardInformationComponent', () => {
...
-            expect(getCourseWithInterestingExercisesForTutorsStub).toHaveBeenCalledTimes(1);
-            expect(getStatsForTutorsStub).toHaveBeenCalledTimes(1);
-        });
+    it('should init component correctly for exam', fakeAsync(() => {
+        comp.ngOnInit();
+        tick();
+
+        expect(comp.isExamMode).toBe(true);
+        expect(comp.courseId).toEqual(10);
+        expect(comp.examId).toEqual(20);
```
Comment (julian-christl) [inline src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts]:
```suggestion
        expect(comp.courseId).toBe(10);
        expect(comp.examId).toBe(20);
```

File: src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts [hunk 3]
```diff
@@ -143,85 +146,129 @@ describe('AssessmentDashboardInformationComponent', () => {
...
+        expect(comp.isTestRun).toBe(false);
+        expect(getExamWithInterestingExercisesForAssessmentDashboardStub).toHaveBeenCalledOnce();
+        expect(getStatsForExamAssessmentDashboardStub).toHaveBeenCalledOnce();
```
Comment (julian-christl) [inline src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts]:
```suggestion
        expect(getExamWithInterestingExercisesForAssessmentDashboardStub).toHaveBeenCalledTimes(1);
        expect(getStatsForExamAssessmentDashboardStub).toHaveBeenCalledTimes(1);
```

File: src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts [hunk 4]
```diff
@@ -143,85 +146,129 @@ describe('AssessmentDashboardInformationComponent', () => {
...
+        } as any as ActivatedRoute;
+        const activatedRoute: ActivatedRoute = fixture.debugElement.injector.get(ActivatedRoute);
+        activatedRoute.snapshot = newRoute.snapshot;
+        TestBed.inject(ActivatedRoute);
+
+        comp.ngOnInit();
+        tick();
+
+        expect(comp.isExamMode).toBe(false);
+        expect(comp.courseId).toEqual(10);
```
Comment (julian-christl) [inline src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts]:
```suggestion
        expect(comp.courseId).toBe(10);
```

File: src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts [hunk 5]
```diff
@@ -143,85 +146,129 @@ describe('AssessmentDashboardInformationComponent', () => {
...
+        expect(comp.examId).toBe(0);
+        expect(getCourseWithInterestingExercisesForTutorsStub).toHaveBeenCalledOnce();
+        expect(getStatsForTutorsStub).toHaveBeenCalledOnce();
```
Comment (julian-christl) [inline src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts]:
```suggestion
        expect(getCourseWithInterestingExercisesForTutorsStub).toHaveBeenCalledTimes(1);
        expect(getStatsForTutorsStub).toHaveBeenCalledTimes(1);
```

File: src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts [hunk 6]
```diff
@@ -143,85 +146,129 @@ describe('AssessmentDashboardInformationComponent', () => {
...
+        comp.updateExercises();
+        expect(comp.exercises).toEqual([textExercise]);
+    });
+
+    it('should toggle showing finished exercises', () => {
+        const updateExercisesSpy = jest.spyOn(comp, 'updateExercises');
+        comp.showFinishedExercises = false;
+        comp.triggerFinishedExercises();
+        expect(comp.showFinishedExercises).toBe(true);
+        expect(updateExercisesSpy).toHaveBeenCalledOnce();
```
Comment (julian-christl) [inline src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts:259]:
```suggestion
        expect(updateExercisesSpy).toHaveBeenCalledTimes(1);
```

File: src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts [hunk 7]
```diff
@@ -143,85 +146,129 @@ describe('AssessmentDashboardInformationComponent', () => {
...
+
+    it('should sort rows', () => {
+        const sortServiceSpy = jest.spyOn(sortService, 'sortByProperty');
+        comp.exercises = [textExercise];
+        comp.exercisesSortingPredicate = 'assessmentDueDate';
+        comp.exercisesReverseOrder = false;
+
+        comp.sortRows();
+
+        expect(sortServiceSpy).toHaveBeenCalledOnce();
```
Comment (julian-christl) [inline src/test/javascript/spec/component/assessment-dashboard/assessment-dashboard.component.spec.ts:270]:
```suggestion
        expect(sortServiceSpy).toHaveBeenCalledTimes(1);
```

**Review by julian-christl — CHANGES_REQUESTED** [review]
Found some test guideline issues, sorry :/ 😅

**Comment by Olimbia23** [comment]
I addressed your suggestions @julian-christl, even though these changes were not introduced by me and were already in develop.

**Review by julian-christl — APPROVED** [review]
Thank you @Olimbia23, appreciate it.

**Review by lschlesinger — DISMISSED** [review]
Re-approve, changes to the test files look good!

**Review by MerlinMehmed — APPROVED** [review]
Retested on TS1, still works as expected!

**Review by julian-christl — APPROVED** [review]
Reapprove Code

**Review by lschlesinger — APPROVED** [review]
Re-approve! Did not see 17 successful github checks for a long time 😄

**Pull request merged**